Mipmap

AlexanderVeselov opened the pull request.

Commit:
Implemented scale_x and scale_y opencl kernels

Commit:
raw pointer replaced with unique_ptr

Commit:
added suport of loading mipmap in Oiio class

Commit:
SceneGraph/Texture interface has been changed

Commit:
Updated Texture struct with mipmap info

Commit:
Added MipmapBuilder interface

Commit:
Updated level scaling kernels

Commit:
Updated 1 chanel scale kernels, added 4 chanel kernels but for them necessary further work

Commit:
Added working 4 chanel scaling kernels

Commit:
Update for payload.cl to fix runtime error due to incorrect offsets

Commit:
Changed SceneGraph/Texture interface

Commit:
Added mipmap class,modified kernels

Commit:
Implemented new version of cl_mipmap

Commit:
Changing SceneGraph and controllers interfaces for mipmap supporting

Commit:
Changed SG::Texture interface, added mipap generation code in clw_scene_controller, changed mipmap builder interface

Commit:
Fixed mipmap behaviour when there's no texture marked as mipmap

Commit:
Fixed crush with generation of mipmap levels

Commit:
Changed mipmap buffer allocation logic in clw_controller. Deleted unnecessary variables from cl_mipmap.*

Commit:
Updated mipmap level scaler kernels, fixed incoorect thread number for scale weights

Commit:
Implemented macros kernel factories in mipmap_level_scaler.cl

Commit:
Refactoring mipmap_level_scaler kernels

Commit:
Removed unecessary variable

Commit:
Fixed scale y kernel. Added mipmap support in image_io interface. Added first mipmap test (Not finished)

Commit:
Added unit tests for mipmap

Commit:
commited half16 unit test

Commit:
merge with master

Commit:
Fixed rpr build error

Commit:
fixed build errors

Commit:
Added auxiliary rays computation for cameras

Commit:
Implemented auxiliary rays partialderivatives computation

Commit:
added mipmap support in texture.cl (OpenCL not compiled)

Commit:
Fixed OpenCL build with mipmap

Commit:
Fixed 8 bit test, added mipmap arg to shade kernel, remooved commited code

Commit:
Added additional unit test for mipmap

Commit:
Merge

AlexanderVeselov committed:
Merge branch 'master' into mipmap

AlexanderVeselov committed:
Clean and format code

AlexanderVeselov committed:
Fix sceen-space derivative calculation
o Fix derivative calculation
o Add tangent vectors calculation based on uv
o Format code

AlexanderVeselov committed:
Refactor mipmap generation

AlexanderVeselov committed:
Add trilinear filtering

o Format code
o Has magic numbers (for now)
o Need to test...

AlexanderVeselov committed:
Fix mipmap seams

o Fix tiling seams
o Fix texel offsets of higher mip levels

AlexanderVeselov committed:
Fix screen space uv derivative calculation

- Remove normalization of dpdu and dpdv vectors
o Refactor kernel code

AlexanderVeselov committed:
Add aux rays and mipmapping to shade kernel

AlexanderVeselov committed:
Solve the issue with tangent space vectors

o dpdu and dpdv vectors in the structure DifferentialGeometry
treat as tangent and bitangent vectors in the kernels.
Actually there is a difference: dpdu/dpdv are need to be
oriented along increasing of uv coordinates, may be not
orthogonal and their length may be not equal to one.

The solution is:
1. Compute dpdu/dpdv in the form, as it need to be used in
screen space uv derivative calculation;
2. Calculate screen space uv derivative;
3. Replace dpdu/dpdv values to the tangent/bitangent values.

AlexanderVeselov committed:
Implement bumpmap texture mipmapping

AlexanderVeselov committed:
Fix texture wrapping where coords are negative

o Replace mod to custom function

AlexanderVeselov committed:
Replace dpdu/dpdv to tangent/bitangent where it's needed

+ Add tangent space calculation with respect to texcoords
o Move screen space UV derivative calculation to the other function

AlexanderVeselov committed:
Add mipmap lookup for specular reflection

AlexanderVeselov committed:
Add mipmap lookup for specular refraction

AlexanderVeselov committed:
Format camera rays generation kernels

AlexanderVeselov committed:
A lot of small fixes

o Fix inconsistent comments
o Fix background aov when plain image has set
- Remove spaces
o Fix texture wrapping in mipmap_level_scaler kernels
o A bit of optimization in texture sampling
- Remove unused code in texture.cl
o Change license year in new files
o Refactor mipmap tests

AlexanderVeselov committed:
Fix Ubuntu build

AlexanderVeselov committed:
Add mipmap level lookup tests

AlexanderVeselov committed:
Merge branch 'master' into mipmap

Kvalme commented on Baikal/Controllers/clw_scene_controller.cpp:
```diff
@@ -841,76 +837,124 @@ namespace Baikal
             auto tex = tex_iter->ItemAs<Texture>();
 
-            WriteTexture(*tex, tex_data_buffer_size, textures + num_textures_written);
+            WriteTexture(*tex, textures + num_textures_written, (int)mip_levels_buffer_size);
```
static_cast please

AlexanderVeselov replied:
Changed parameter type to std::size_t

Kvalme commented on Baikal/Controllers/clw_scene_controller.cpp:
```diff
@@ -841,76 +837,124 @@ namespace Baikal
+
+            for (std::size_t i = 0; i < tex->GetMipLevelCount(); ++i)
+            {
+                WriteMipLevel(*tex, mip_levels + num_mip_levels_written, i, (int)texturedata_buffer_size);
```
static_cast please

AlexanderVeselov replied:
Changed parameter type to std::size_t

Kvalme commented on Baikal/Controllers/clw_scene_controller.cpp:
```diff
@@ -1214,23 +1258,31 @@ namespace Baikal
     }
 
-    void ClwSceneController::WriteTexture(Texture const& texture, std::size_t data_offset, void* data) const
+    void ClwSceneController::WriteTexture(Texture const& texture, void* data, int mip_levels_offset) const
```
std:;size_t instead of int?

AlexanderVeselov replied:
Ok, you're right, it will improve encapsulation.

Kvalme commented on Baikal/Controllers/clw_scene_controller.h:
```diff
@@ -95,9 +93,13 @@ namespace Baikal
         void UpdateIntersector(Scene1 const& scene, ClwScene& out) const;
         void UpdateIntersectorTransforms(Scene1 const& scene, ClwScene& out) const;
+        // Clear intersector and load meshes into it.
+        void ReloadIntersector(Scene1 const& scene, ClwScene& inout) const;
```
What is the reason to move it to private from protected?

AlexanderVeselov replied:
Because of better encapsulation. And we don't use protection-features in here

Kvalme commented on Baikal/Kernels/CL/bxdf_uberv2.cl:
```diff
@@ -70,9 +70,9 @@ void UberV2_ApplyShadingNormal(
-        dg->n = normalize(shader_data->shading_normal.z * dg->n + shader_data->shading_normal.x * dg->dpdu + shader_data->shading_normal.y * dg->dpdv);
-        dg->dpdv = normalize(cross(dg->n, dg->dpdu));
-        dg->dpdu = normalize(cross(dg->dpdv, dg->n));
+        dg->n = normalize(shader_data->shading_normal.z * dg->n + shader_data->shading_normal.x * dg->tangent + shader_data->shading_normal.y * dg->bitangent);
```
add comment Why dpdu/dpdv was replaced with tangent/bitangent

AlexanderVeselov replied:
I've added a note about this in Scene_FillDifferentialGeometry

Kvalme commented on Baikal/Kernels/CL/common.cl:
```diff
@@ -69,4 +69,6 @@ THE SOFTWARE.
 #define VISIBILITY_MASK_BOUNCE(i) (VISIBILITY_MASK_PRIMARY << (i))
 #define VISIBILITY_MASK_BOUNCE_SHADOW(i) (VISIBILITY_MASK_SHADOW << (i))
 
+#define NULL (0)
```
You don't need it

AlexanderVeselov replied:
Fixed

Kvalme commented on Baikal/Kernels/CL/integrator_bdpt.cl:
```diff
@@ -180,6 +180,10 @@ KERNEL void GenerateLightVertices(
     // Ray batch
     GLOBAL ray const* rays,
+    // auxiliary ray in x dimension
+    GLOBAL aux_ray const* x_aux_rays,
```
seems to be unused in this kernel

AlexanderVeselov replied:
Fixed

Kvalme commented on Baikal/Kernels/CL/mipmap_level_scaler.cl:
```diff
@@ -0,0 +1,261 @@
+#pragma OPENCL EXTENSION cl_khr_fp16 : enable
+
+KERNEL
+void ComputeWeights_NoRounding(
```
add comments to all kernels

Kvalme replied:
and all functions in this file

AlexanderVeselov replied:
Added

Kvalme commented on Baikal/Kernels/CL/mipmap_level_scaler.cl:
```diff
@@ -0,0 +1,261 @@
+
+    if (id < size)
+    {
+        weights[id].x = .0f;
```
why you need this kernel?

Kvalme replied:
also weights[id] = (float3)(0.f,  0.5f, 0.5f);
Why 0.f for x and 0.5f for y?

AlexanderVeselov replied:
The kernel is used when size (width/height) of current mip level can be divided by two without remainder (is even).

In general the convolution formula is:
> next_level[x] = w0 * curr_level[2 * x - 1] +  w1 * curr_level[2 * x] + w2 * curr_level[2 * x + 1]

x, y, z components of weights buffer correspond to w0, w1, w2 weights.
When the current mip level size is even, we apply simpler version:
> next_level[x] = 0.5 * curr_level[2 * x] + 0.5 *curr_level[2 * x + 1]

But if the size of mip level is odd, we divide it by two and round up. In this case the convolution weights are:
> w0 = (size - x - 1) / (2 * n - 1)
w1 = size / (2 * n - 1)
w2 = x / (2 * n - 1)

As we can see, in the first case the x component is not being used, so we set it to zero.

Kvalme commented on Baikal/Kernels/CL/mipmap_level_scaler.cl:
```diff
@@ -0,0 +1,261 @@
+{
+    int id = get_global_id(0);
+
+    float denominator = 2.f * size - 1.f;
```
what if size equal to 0.5?

AlexanderVeselov replied:
But this is an integer...

Kvalme commented on Baikal/Kernels/CL/mipmap_level_scaler.cl:
```diff
@@ -0,0 +1,261 @@
+    if (id == 0)
+    {
+        weights[id].x = .0f;
+        weights[id].y = ((float)size) / denominator;
```
are you ready for NaNs?

AlexanderVeselov replied:
### *OH SHI—!*

AlexanderVeselov replied:
Actually, there shouldn't be any NaNs, since size should be >= 1 and denominator never equals zero.

Kvalme commented on Baikal/Kernels/CL/mipmap_level_scaler.cl:
```diff
@@ -0,0 +1,261 @@
+
+// Convert uchar4/half4/float4 type to user and store it in buffer by index
+inline void SetValue_uchar4(
+    GLOBAL uchar4* buf, int index, float4 value)
```
comments to all arguments

Kvalme replied:
see bxdf_uberv2.cl for comment style

AlexanderVeselov replied:
Fixed

Kvalme commented on Baikal/Kernels/CL/mipmap_level_scaler.cl:
```diff
@@ -0,0 +1,261 @@
+    buf[index] = value;
+}
+
+// Level scaler kernels scheme
```
??? Comment should answer on question what this function/macros do.
Please, write detailed comments especially when you add mactos with hell of a code inside

AlexanderVeselov replied:
Fixed

Kvalme commented on Baikal/Kernels/CL/monte_carlo_renderer.cl:
```diff
@@ -37,9 +37,9 @@ THE SOFTWARE.
 // This kernel is being used if aperture value = 0.
 KERNEL
-void PerspectiveCamera_GeneratePaths(
+void PerspectiveCamera_GenerateRays(
```
You reordered this functions in file?

AlexanderVeselov replied:
No, I have renamed them from "paths" to "rays", because I think it's more actual naming

Kvalme replied:
Please, revert it then.

AlexanderVeselov replied:
Fixed

Kvalme commented on Baikal/Kernels/CL/payload.cl:
```diff
@@ -215,56 +215,73 @@ enum TextureFormat
+    float3 dpdx;
+    float3 dpdy;
+    // UV derivatives with respect to screen coords
+    float2 duvdx;
```
doesn't it possible to convert them to each other?

AlexanderVeselov replied:
I don't understand. What to what?

Kvalme commented on Baikal/SceneGraph/texture.cpp:
```diff
@@ -62,29 +58,127 @@ namespace Baikal
+
+        m_mip_sizes = mip_sizes;
+
+        std::transform(m_mip_sizes.begin(), m_mip_sizes.end(), m_mip_sizes.begin(),
```
What are you doing here?

AlexanderVeselov replied:
In the previous version std::max(1, size.z) has been used to prevent a user setting z-size to 0

Kvalme replied:
there are only 2 places where user can set size. So please check size at set-time

AlexanderVeselov replied:
Ok, I've done this

AlexanderVeselov committed:
Fix issues from the review №1

+ Add MORE comments
+ Add note about changing from dpdu/dpdv to tangent/bitangent vectors
o Change type cast in ClwSceneController::WriteTexture and WriteMipLevel
o Restore camera ray generation kernel names
o Fix zero texture depth in Rpr layer
- Remove std::transform from Baikal::Texture constructors

Kvalme commented on Baikal/Kernels/CL/scene.cl:
```diff
@@ -247,53 +265,192 @@ void Scene_FillDifferentialGeometry(// Scene
+    float2 duv02 = uv0 - uv2;
+    float2 duv12 = uv1 - uv2;
+    float  det = duv02.x * duv12.y - duv02.y * duv12.x;
+    bool   degenerate_uv = fabs(det) < 1e-08f;
```
please, add constant for this value (FLT_EPSILON for example) and use it.

AlexanderVeselov replied:
Done

Kvalme commented on Baikal/Kernels/CL/scene.cl:
```diff
@@ -247,53 +265,192 @@ void Scene_FillDifferentialGeometry(// Scene
-}
 
+    // Initialize screen space uv derivatives
+    diffgeo->duvdx = diffgeo->duvdy = 0.0f;
```
split into two lines.

AlexanderVeselov replied:
Fixed

Kvalme commented on Baikal/Kernels/CL/scene.cl:
```diff
@@ -247,53 +265,192 @@ void Scene_FillDifferentialGeometry(// Scene
+    // Bitangent vector need to be orthogonal to tangent, since we use
+    // the advantage of an orthogonal matrix that is easily invertible
+    diffgeo->bitangent = normalize(cross(diffgeo->tangent, diffgeo->n));
+/*
```
remove dead code

AlexanderVeselov replied:
I've uncommented it

Kvalme commented on Baikal/Kernels/CL/scene.cl:
```diff
@@ -247,53 +265,192 @@ void Scene_FillDifferentialGeometry(// Scene
+                          float2* uv_derivative
+                          )
+{
+    //float3 o = vload_half3(0, (GLOBAL half*)&my_ray->o);
```
remove dead code

AlexanderVeselov replied:
Fixed

Kvalme commented on Baikal/Kernels/CL/scene.cl:
```diff
@@ -247,53 +265,192 @@ void Scene_FillDifferentialGeometry(// Scene
+
+    float3 p = o + d * t;
+
+    // Next we need to find uv offset from the position offset
```
remove dead code

AlexanderVeselov replied:
This is the comment with the formula, not a line of code

Kvalme commented on Baikal/Kernels/CL/scene.cl:
```diff
@@ -247,53 +265,192 @@ void Scene_FillDifferentialGeometry(// Scene
+    }
+
+    // Solve linear system using Cramer rule
+    float matrix_det = derivative_matrix.x * derivative_matrix.w - derivative_matrix.y * derivative_matrix.z;
```
can it be 0?

AlexanderVeselov replied:
Yes, checked

Kvalme commented on Baikal/Kernels/CL/scene.cl:
```diff
@@ -247,53 +265,192 @@ void Scene_FillDifferentialGeometry(// Scene
+    float det2 = derivative_matrix.x * delta_p.y - derivative_matrix.z * delta_p.x;
+
+    *position_derivative = p - diffgeo->p;
+    *uv_derivative = make_float2(det1, det2) / matrix_det;
```
NaNs?

AlexanderVeselov replied:
Added denominator checking

AlexanderVeselov committed:
Fix issues from the review №2

- Remove commented code
o Fix bug with zero tangent vector when a triangle has no uvs
+ Add a comment about singular reflected/refracted rays
+ Add zero denominator check

Kvalme commented:
Ok. Create external CR

AlexanderVeselov committed:
Merge branch 'master' into mipmap

AlexanderVeselov committed:
Fix Intel HD driver crash on OSX machine

- Remove zero-pointer check
